README.md
# Introduction 
Ce repos contient les sources AL de l'App CallioPack CAGTX Tax Management (Gestion des taxes).

# Pour commencer à développer 
1. Prérequis systèmes et logiciels , et paramétrage environnement de développement local AL : [Business Central-AL Configuration générale](https://3lionline.sharepoint.com/teams/projectsv2/ressources/Documents%20administration/Business%20Central-AL%20Configuration%20g%C3%A9n%C3%A9rale.pptx?d=wae4701a7eee64c068e0e80c285905e01)
2. Création de son environnement de développement projet​ : [Business Central-AL Environnement de dev projet - Sandbox Docker](https://3lionline.sharepoint.com/teams/projectsv2/ressources/Documents%20administration/Business%20Central-AL%20Environnement%20de%20dev%20projet%20-%20Sandbox%20Docker.pptx?d=wf690e31ab7654262a3249dc5eaf112f3)
3. Normes de développement : [Normes de développement NAV et Business Central](https://3lionline.sharepoint.com/teams/projectsv2/ressources/SitePages/Normes%20de%20d%C3%A9veloppement%20NAV%20et%20Business%20Central.aspx) 

# Build & Test

# Contribuer


scripts/ReleaseNotes.txt
v4.1
#5218 : 21/01/2022 : PMN : [BreakingChange] Build AL : Publish ChangeNote to ArtifactFolder (ArtifactType subfolder)
#5218 : 20/01/2022 : PMN : Build AL : Fix build/hotfix number and contextSensitiveHelpUrl not updated in app.json if Convert main App (renumbering objects, renaming prefixes, new App Id)
#5218 : 20/01/2022 : PMN : Build AL : Fix ProjectSettings.xml update if don't exist nodes DevLocalInstallScripts and BuildLocalInstallScripts
#5218 : 17/01/2022 : PMN : Deploy AL : Add new depencencies management (subfolder of ArtifactType)
#4680 : 17/01/2022 : PMN : [Hotfix] Dev AL : Rollback set Docker container default isolation to process (bugs fixed in BCContainerHelper 2.0.17)
#4679 : 25/11/2021 : PMN : Build AL : Modify pattern for RapidStart Packages object ids renumbering (renumber now all contents)
#4581 : 22/11/2021 : PMN : Build AL : Add update RapidStart Packages object name prefix
#4583 : 22/11/2021 : PMN : Build AL : Add renumber also CaptionClass with 3 values separated with , (2nd number = table no to renumber, 3rd number = field no to renumber)
#4474 : 03/11/2021 : PMN : [BreakingChange] Build/Deploy AL : Add dependencies management deployment with multiple artifact types (specially for dependencies between CallioPacks)
                           - Params $MainAppId and $MainAppPrefix for NAVALRunBuild.ps1 are mandatory for dependencies management deployment
#3506 : 25/10/2021 : PMN : Build/ Deploy AL/CAL : Fix improve error return for DevOps (Errors not catched as Exception)
#3480 : 23/10/2021 : PMN : Build/Deploy AL : Add dependencies management with multiple artifact types (specially for dependencies between CallioPacks)

v4.0
#3506 : 20/10/2021 : PMN : Build/ Deploy AL/CAL : Improve error return for DevOps
#4223 : 18/10/2021 : PMN : Deploy AL (Git) : Add --depth 1 to Invoke-GitCloneRemoteRepos and add text to email to merge scripts to other branches
#4224 : 18/10/2021 : PMN : Dev AL : Set Docker container Dns to 8.8.8.8
#4254 : 18/10/2021 : PMN : Dev AL : Set Docker container default isolation to hyperv

v3.9
#2433 : 22/07/2021 : PMN : Fix compilation error not catched as build error
#3313 : 22/07/2021 : PMN : Build CAL : Fix MenuSuite compilation if only one line in the log file
#2823 : 30/06/2020 : PMN : Build Help : Manage toc.yml files in subfolders
#2821 : 29/06/2021 : PMN : Build AL : Fix Git Change Note FromDate format (MM/dd/yyyy) 
#2408 : 23/06/2021 : PMN : Build AL : Fix multi-range management from App renumbering in xliff files
#2408 : 16/06/2021 : PMN : Build AL : Add multi-range management from App renumbering (comma separated list for param NewMainAppIdRangeFrom )
#2495 : 26/05/2021 : PMN : Deploy AL/CAL (Mail) : Fix attachments with the same image
#2138 : 04/05/2021 : PMN : Deploy AL (Git) : Fix ProjectSettings.xml update if exists several Apps with a similar name
#2315 : 03/05/2021 : PMN : Deploy AL (OnPrem) : Add a license file to use param (especially for to deploy a BaseApp with a Dev license)
#1360 : 15/03/2021 : PMN : Dev AL : Fix posh-git new version installation
#1312 : 10/03/2021 : PMN : Deploy AL (OnPrem) : Fix error reinstall dependencies uninstalled Apps with other version already installed (caused by ForceSync)
#1250 : 05/03/2021 : PMN : Deploy AL (OnPrem) : Fix error uninstall App with ForceSync
#430 : 02/03/2021 : PMN : Dev AL : Migration to BcContainerHelper
#421 : 23/02/2021 : PMN : Build AL : Add RapidStart Packages renumbering in Convert-NavApp, Artifact publishing and deploy to Git repos
#1169 : 22/02/2021 : PMN : Build CAL : Fix license expiration warning
#622 : 19/02/2021 : PMN : Deploy AL (OnPrem) : Fix error when publish/install a previous version or a same version with ForceSync
#1152 : 19/02/2021 : PMN : Build CAL : Fix SQL Server name and NAV Server name wrong display, MaxTry default value changed from 3 to 1
#169 : 17/02/2021 : PMN : Build CAL : Fix Docker images removed (Add BC artifacts management (BC Docker images replacement))
#422 : 12/02/2021 : PMN : Build AL : Add publish Change Note
#432 : 04/01/2021 : PMN : Deploy AL (Azure VM) : Add Azure Service Principal authentication with secret
#429 : 30/12/2020 : PMN : Build AL : Fix Sign-NavContainerApp (https://github.com/microsoft/navcontainerhelper/issues/1579)
#424 : 23/12/2020 : PMN : Build Help : Adaptation for all solutions (not only CallioPacks Business Central)
#417 : 22/12/2020 : PMN : Build/Deploy AL : Fix deployment failed if no dependent Apps (add empty apps.json in Dependencies folder)
#359 : 07/12/2020 : PMN : Deploy AL : Fix deploy PerTenant to Git repos with ArtifactType PerTenant
#317 : 04/12/2020 : PMN : Deploy AL : Simplify OnPrem App deployment with Depencencies

v3.8
#315 : 27/11/2020 : PMN : Deploy AL : Add Azure VM App deployment with Depencencies and auto-starting/auto-stopping VM if not started
#307 : 27/11/2020 : PMN : Build AL : Add apps.json to Dependencies artifact folder to simplify dependent Apps deployment

v3.7
#259 : 15/11/2020 : PMN : Build/Dev AL : Adaptations for BC17 BaseApp (MS Application App cannot be unpublished and MS Base Application App cannot be renammed)
#258 : 15/11/2020 : PMN : Build/Dev AL : Fix (workarround) NavContainetHelper issue in New-NavContainer with -includeTestToolkit for containers BC >= 17.2

v3.6
#86 : 12/11/2020 : PMN : Build AL/CAL : Add clean artifacts cache and remove containers (NAVCleanBuild script)
#255 : 12/11/2020 : PMN : Build/Dev AL : BcContainerHelper migration preparation
#256 : 12/11/2020 : PMN : Dev AL : Fix encoding issue with PowerShell (remove accents in folder and file names)

v3.5
#253 : 11/11/2020 : PMN : Build/Dev AL : Fix (workarround) NavContainetHelper issue in New-NavContainer with -includeTestToolkit for BC18 containers
#249 : 06/11/2020 : PMN : Dev AL : Add LS HardwareStation shortcuts (for projects based on LS)

v3.4
#243 : 02/11/2020 : PMN : Build AL : Add dependent Apps (based on BuildLocalInstallScripts node) to Dependencies artifact folder (for future deployment)

v3.3

scripts/SetupMsi.ps1
<#
------------------------------------------------------------------------------------------------------------------
Authors      : PMN (pmoison@3li.com)
Copyright    : 3Li Business Solutions
Description  : Run Msi
------------------------------------------------------------------------------------------------------------------
References   :
Dependencies : 
------------------------------------------------------------------------------------------------------------------
Revisions    : 17/01/2019 : PMN : Initial version
------------------------------------------------------------------------------------------------------------------
#>
Param(
  [string]$MsiFilename = "", # Msi filename
  [string]$MsiOption = "", # Msi options
  [string]$MsiParameters = "" # Msi parameters
)

# Run the Msi installer 
$Arguments = "$MsiOption ""$MsiFilename"" $MsiParameters"
Write-Host "Running Msi ($MsiFilename) with option ($MsiOption) and parameters ($MsiParameters)"
$App = Start-Process "msiexec.exe" -Wait -ArgumentList $Arguments


scripts/SetupNAVLSSetup.ps1
<#
------------------------------------------------------------------------------------------------------------------
Authors      : PMN (pmoison@3li.com)
Copyright    : 3Li Business Solutions
Description  : Run LS Retail Setups
------------------------------------------------------------------------------------------------------------------
References   :
Dependencies : 
------------------------------------------------------------------------------------------------------------------
Revisions    : 10/08/2018 : PMN : Initial version
               13/08/2018 : PMN : Fix bug space before /TASKS
               14/11/2018 : PMN : Add param Folder for destination folder
               #249 : 06/11/2020 : PMN : Add LS HardwareStation shortcuts (and new param $ShortcutsFolder)
------------------------------------------------------------------------------------------------------------------
#>
Param(
  [string]$LSSetupExe = "", # LS Retail setup exe
  [string]$Tasks = "", # Setup task list
  [string]$Folder = "", # Setup default folder
  [string]$ShortcutsFolder="" # destination folder for shortcuts
)

$Arguments = "/SILENT /NOCANCEL"
if ($Tasks -ne "")
{
	$Arguments += " /TASKS=""$Tasks"""
}
if ($Folder -ne "")
{
	$Arguments += " /DIR=""$Folder"""
}

# Run the LS Retail installer 
Write-Host "Running LS Retail Setup ($LSSetupExe) with arguments ($Arguments)"
$App = Start-Process $LSSetupExe -Wait -ArgumentList $Arguments

# Create shortcuts
if ($ShortcutsFolder -ne "") {
  $TargetFolder = "C:\Program Files (x86)\LS Retail\LSHardwareStation"
  if ($Folder -ne "") { $TargetFolder = $Folder }
  if (Test-path $TargetFolder ) {
    # Create LS HardwareStation - Start shortcut
		$Shell = New-Object -ComObject ("WScript.Shell")
		$ShortCutFile = "$ShortcutsFolder\LS HardwareStation - Start.lnk"
		$ShortCut = $Shell.CreateShortcut("$ShortCutFile")
		$ShortCut.TargetPath = "$TargetFolder\LSHardwareStation.exe"
		$ShortCut.Arguments = "-debug"
		$ShortCut.WorkingDirectory = "$TargetFolder\"
		$ShortCut.WindowStyle = 1
		$ShortCut.Save()
		$bytes = [System.IO.File]::ReadAllBytes($ShortCutFile)
		$bytes[0x15] = $bytes[0x15] -bor 0x20 #set byte 21 (0x15) bit 6 (0x20) ON
    [System.IO.File]::WriteAllBytes($ShortCutFile, $bytes)
    
    # Create LSVirtualStation - Start shortcut
		$Shell = New-Object -ComObject ("WScript.Shell")
		$ShortCutFile = "$ShortcutsFolder\LS VirtualStation - Start.lnk"
		$ShortCut = $Shell.CreateShortcut("$ShortCutFile")
		$ShortCut.TargetPath = "$TargetFolder\LSVirtualStation.exe"
		$ShortCut.Arguments = ""
		$ShortCut.WorkingDirectory = "$TargetFolder\"
		$ShortCut.WindowStyle = 1
		$ShortCut.Save()
		$bytes = [System.IO.File]::ReadAllBytes($ShortCutFile)
		$bytes[0x15] = $bytes[0x15] -bor 0x20 #set byte 21 (0x15) bit 6 (0x20) ON
    [System.IO.File]::WriteAllBytes($ShortCutFile, $bytes)
    
    # Create LS HardwareStation - Management Portal shortcut
    $Shell = New-Object -ComObject ("WScript.Shell")
    $ShortCutFile = "$ShortcutsFolder\LS HardwareStation - Management Portal.url"
    $ShortCut = $Shell.CreateShortcut("$ShortCutFile")
    $ShortCut.TargetPath = "http://localhost:8088/"
    $ShortCut.Save()
  
    # Create LS HardwareStation - Documentation shortcut
    $Shell = New-Object -ComObject ("WScript.Shell")
    $ShortCutFile = "$ShortcutsFolder\LS HardwareStation - Documentation.url"
    $ShortCut = $Shell.CreateShortcut("$ShortCutFile")
    $ShortCut.TargetPath = "https://help.lscentral.lsretail.com/Content/LS%20Retail/POS/Hardware/LS%20Hardware%20Station.htm"
    $ShortCut.Save()

  }
} 




scripts/SetupNAVAppContainer.ps1
<#
------------------------------------------------------------------------------------------------------------------
Authors      : PMN (pmoison@3li.com)
Copyright    : Calliope 3Li
Description  : Install NAV App in container
------------------------------------------------------------------------------------------------------------------
References   :
Dependencies : 
------------------------------------------------------------------------------------------------------------------
Revisions    : 26/02/2020 : PMN : Initial version
               #172 : 22/09/2020 : PMN : Add skipverification
               #212 : 14/10/2020 : PMN : Fix install in container new version of App, BaseApp and reinstall uninstalled Apps
               #243 : 02/11/2020 : PMN : Add copy App to $CopyToFolder folder
               #307 : 27/11/2020 : PMN : Add apps.json to folder $CopyToFolder to simplify dependent Apps deployment
               #3480 : 23/10/2021 : PMN : Add params AppSourcePrefix, OnPremCfmdPrefix, PerTenantPrefix (for dependencies management with multiple artifact types)
------------------------------------------------------------------------------------------------------------------
#>
Param(
  [string]$ContainerName = "", # Container name
  [string]$AppFile = "", # App file
  [string]$AppName = "", # App name
  [string]$SyncMode = "Add", # Sync mode
  [string]$CopyToFolder = "", # foldder to copy App
  [string]$AppSourcePrefix = "", # AppSource App object prefix
  [string]$OnPremCfmdPrefix = "", # OnPremCfmdPrefix App object prefix
  [string]$PerTenantPrefix = "" # PerTenantPrefix App object prefix
)
# Init 
$ErrorActionPreference = "Stop"

# Get App file info
$ContainerAppFile = "c:\run\my\$(Split-Path $AppFile -Leaf)"
$ContainerId = Get-NavContainerId -containerName $ContainerName
$Session = New-PSSession -ContainerId $ContainerId -RunAsAdministrator
Copy-Item $AppFile -Destination $ContainerAppFile -ToSession $Session -Force
Remove-PSSession $Session
$AppFileInfo = Invoke-ScriptInNavContainer -containerName $ContainerName -argumentList $ContainerAppFile -scriptblock {Param($ContainerAppFile)
    Get-NavAppInfo -Path "$ContainerAppFile"
    Remove-Item "$ContainerAppFile"
}

# Copy App to $CopyToFolder folder
if ($CopyToFolder -ne "")
{
    # Copy App 
    Write-Host "Copy $AppFile to $CopyToFolder folder ..."
    New-Item -Path "$CopyToFolder" -ItemType Directory -Force | Out-Null
    Copy-Item -Path "$AppFile" -Destination "$CopyToFolder\$(Split-Path $AppFile -leaf)" -Force

    # Update apps.json
    $Apps = @()
    if (Test-Path "$CopyToFolder\apps.json") {
        $JsonApps = Get-Content "$CopyToFolder\apps.json" | ConvertFrom-Json 
        foreach($JsonApp in $JsonApps) {
            if (-not (($JsonApp.publisher -eq $AppFileInfo.Publisher -and $JsonApp.name -eq $AppFileInfo.Name) -or ($JsonApp.appId -eq $AppFileInfo.AppId))) {
                $Apps += [pscustomobject]@{fileName=$JsonApp.fileName;appId=$JsonApp.appId;publisher=$JsonApp.publisher;name=$JsonApp.name;version=$JsonApp.version;brief=$JsonApp.brief;appSourcePrefix=$JsonApp.appSourcePrefix;onPremCfmdPrefix=$JsonApp.onPremCfmdPrefix;perTenantPrefix=$JsonApp.perTenantPrefix }
            } 
        }
    }
    $Apps += [pscustomobject]@{fileName=$(Split-Path $AppFile -Leaf);appId=$AppFileInfo.AppId;publisher=$AppFileInfo.Publisher;name=$AppFileInfo.Name;version=$AppFileInfo.Version.ToString();brief=$AppFileInfo.Brief;appSourcePrefix=$AppSourcePrefix;onPremCfmdPrefix=$OnPremCfmdPrefix;perTenantPrefix=$PerTenantPrefix}
    $Apps | ConvertTo-Json -Depth 10 | Set-Content "$CopyToFolder\apps.json"

}

# Set AppName
$AppName = $AppFileInfo.Name

# Get installed Apps
$InstalledApps = Get-NavContainerAppInfo -containerName $ContainerName -tenantSpecificProperties -sort DependenciesFirst | Where-Object { $_.IsInstalled }
Write-Host "Installed Apps:"
foreach ($InstalledApp in $InstalledApps) { Write-Host "- $($InstalledApp.Name) Publisher=$($InstalledApp.Publisher) Version=$($InstalledApp.Version)" }

# Publish 
Publish-NavContainerApp -containerName $ContainerName -appFile $AppFile -skipVerification

if ($SyncMode -eq "ForceSync")
{
    # Uninstall
    UnInstall-NavContainerApp -containerName $ContainerName -appName $AppName -Force
}

# Sync 
Sync-NavContainerApp -containerName $ContainerName -appName $AppName -appVersion $AppFileInfo.Version -Mode $SyncMode -Force

# Install
if (Get-NavContainerAppInfo -containerName $ContainerName -tenantSpecificProperties | Where-Object { $_.Name -eq $AppName -and $_.Version -ne $AppFileInfo.Version -and ($_.IsInstalled -or $_.ExtensionDataVersion -ne $AppFileInfo.Version)}) {
    Start-NavContainerAppDataUpgrade -containerName $ContainerName -appName $AppName -appVersion $AppFileInfo.Version
} else {
    Install-NavContainerApp -containerName $ContainerName -appName $AppName -appVersion $AppFileInfo.Version
}

# Unpublish previous versions
Write-Host "Unpublish previous versions of $AppName"
Get-NavContainerAppInfo -containerName $ContainerName -tenantSpecificProperties | Where-Object { -not $_.IsInstalled -and $_.Name -eq $AppName } | ForEach-Object { UnPublish-NavContainerApp -containerName $ContainerName -appName $AppName -version $_.Version }

# Reinstall uninstalled apps
Write-Host "Reinstall uninstalled apps"
$UninstalledApps = Get-NavContainerAppInfo -containerName $ContainerName -tenantSpecificProperties -sort DependenciesFirst | Where-Object { -not $_.IsInstalled -and $_.Name -ne $AppName } 
foreach ($UninstalledApp in $UninstalledApps) {
    $InstalledApp = $InstalledApps | Where-Object { $_.Name -eq $UninstalledApp.Name -and $_.Version -eq $UninstalledApp.Version }
    if ($InstalledApp) { Install-NavContainerApp -containerName $ContainerName -appName $InstalledApp.Name -appVersion $InstalledApp.Version }
}


scripts/GitManagement.ps1
<#
------------------------------------------------------------------------------------------------------------------
Authors      : PMN (pmoison@3li.com)
Copyright    : Calliope 3Li
Description  : Git management
------------------------------------------------------------------------------------------------------------------
References   :
Dependencies : posh-git (https://github.com/dahlbyk/posh-git)
               Git
------------------------------------------------------------------------------------------------------------------
Revisions    : #12633 : 15/04/2020 : PMN : Initial version
               #14111 : 29/07/2020 : PMN : Add functions Invoke-GitCloneRemoteRepos, Invoke-GitCommit, Invoke-GitAdd and Set-GitUserAndCredential
			   #1360 : 15/03/2021 : PMN : Fix posh-git new version installation
			   #4223 : 18/10/2021 : PMN : Add --depth 1 to Invoke-GitCloneRemoteRepos
------------------------------------------------------------------------------------------------------------------
#>

Function Install-PoshGit
{
	$Module = Get-InstalledModule -Name posh-git -ErrorAction SilentlyContinue
	if ($Module) 
	{
		$VersionStr = $Module.Version.ToString()
		Write-Host "posh-git $VersionStr is installed"
		$OnlineModule = Find-Module -Name posh-git -ErrorAction SilentlyContinue
		if ($OnlineModule) 
		{
			$LatestVersion = $OnlineModule.Version 
			$PoshGitVersion = $LatestVersion.ToString()
			Write-Host "posh-git $PoshGitVersion is the latest version"
			if ($PoshGitVersion -ne $Module.Version) 
			{
				Write-Host "Updating posh-git to $PoshGitVersion"
				Update-Module -Name posh-git -Force -RequiredVersion $PoshGitVersion
				Write-Host "posh-git updated"
			}
		} else {
			Write-Warning "Unable to check online latest version of posh-git (check Internet connectivity)"
		}
	} 
	else 
	{
		if (!(Get-PackageProvider -Name NuGet -ListAvailable -ErrorAction SilentlyContinue)) {
			Write-Host "Installing NuGet Package Provider"
			Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.208 -Force -WarningAction SilentlyContinue | Out-Null
		}
		Write-Host "Installing posh-git"
		Install-Module -Name posh-git -Force
		$Module = Get-InstalledModule -Name posh-git -ErrorAction SilentlyContinue
		$VersionStr = $Module.Version.ToString()
		Write-Host "Add current branch to PowerShell prompt"
		Add-PoshGitToProfile -AllHosts -Force
		Write-Host "posh-git $VersionStr installed"
	}
}

Function Get-GitCurrentBranch
{
    Param(
    [Parameter(ValueFromPipelinebyPropertyName=$True)]
    [String]$GitFolder = (Get-Location)
    )
	$env:GIT_REDIRECT_STDERR = '2>&1'
	$CurrentBranch = ""
	$CurrentLocation = Get-Location
	try
	{
		Set-Location -Path $GitFolder
		git branch | foreach {
			if ($_ -match "^\* (.*)") {
				$CurrentBranch += $matches[1]
			}
		}
	}
	finally
	{
		Set-Location -Path $CurrentLocation
	}
	return $CurrentBranch
}

Function Get-GitBranches
{
    Param(
    [Parameter(ValueFromPipelinebyPropertyName=$True)]
    [String]$GitFolder = (Get-Location)
    )
	$env:GIT_REDIRECT_STDERR = '2>&1'
	$Branches = @()
	$CurrentLocation = Get-Location
	try
	{
		Set-Location -Path $GitFolder
		git branch | foreach {
			if ($_ -match "^\* (.*)") {
				$Branches += $matches[1]
			} else { $Branches += $_.Trim()}
		}
	}
	finally
	{
		Set-Location -Path $CurrentLocation
	}
	return $Branches
}

Function Write-GitWelcomeText
{
	Write-host "Welcome to Git PowerShell Prompt" -ForegroundColor Yellow
	Write-host ""
	Write-host "Git commands help" -ForegroundColor Yellow
	git
	Write-host ""
	Write-host "Git additional Powershell commands" -ForegroundColor Yellow
	Write-host "Get-GitBranches               Get Git local branches for a specific local folder"
	Write-host "Get-GitCurrentBranch          Get Git local current branch for a specific local folder"
	Write-host "Switch-GitBranch              Switch Git local current branch for a specific local folder"
	Write-host ""
	Write-host "Git local branches list" -ForegroundColor Yellow
	git branch
	Write-host ""
}

Function Switch-GitBranch
{
    Param(
    [Parameter(ValueFromPipelinebyPropertyName=$True)]
    [String]$GitFolder = (Get-Location),
    [Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$TargetGitBranch
    )
	$GitBranches = Get-GitBranches -GitFolder $GitFolder
	if ($GitBranches.Contains($TargetGitBranch))
	{
		$env:GIT_REDIRECT_STDERR = '2>&1'
		$CurrentLocation = Get-Location
		try
		{
			Set-Location -Path $GitFolder
			git switch $TargetGitBranch
		}
		finally
		{
			Set-Location -Path $CurrentLocation
		}
	}
	else
	{
		Write-Error "Error : $TargetGitBranch in not a valid branch name."
	}
}

Function Get-GitCurrentBranchColor
{
	Param(
    [Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitBranch,
    [Parameter(ValueFromPipelinebyPropertyName=$True)]
    [String]$GitBranchesSupported
    )
	$SplitGitBranchesSupported = $GitBranchesSupported.Split(",")
	$Result = "Red"
	if ($SplitGitBranchesSupported.Contains($GitBranch))
	{
		if ($SplitGitBranchesSupported[0] -eq $GitBranch) { $Result = "Green" } else { $Result = "Yellow" }
	}
	return $Result
}

Function Invoke-GitCloneRemoteRepos
{
    Param(
    [Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitRemoteReposUrl,
    [Parameter(ValueFromPipelinebyPropertyName=$True)]
    [String]$GitBranch="",
    [Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitFolder
    )
	$env:GIT_REDIRECT_STDERR = '2>&1'
	if ($GitBranch -eq "") {
		Write-Host "Git clone defaut branch (with --depth1) from remote repos $GitRemoteReposUrl"
		git clone --depth 1 --single-branch $GitRemoteReposUrl "$GitFolder"
	} else {
		Write-Host "Git clone $GitBranch branch (with --depth1) from remote repos $GitRemoteReposUrl"
		git clone --depth 1 --single-branch --branch $GitBranch $GitRemoteReposUrl "$GitFolder"
	}
}

Function Invoke-GitCommit
{
    Param(
	[Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitFolder,
    [Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitCommitMessage,
	[Parameter(ValueFromPipelinebyPropertyName=$True)]
	[switch]$GitCommitAndPush
	)
	$env:GIT_REDIRECT_STDERR = '2>&1'
	$CurrentLocation = Get-Location
	try
	{
		Set-Location -Path $GitFolder
		git commit -a -m $GitCommitMessage
		if ($GitCommitAndPush) { git push --porcelain }
	}
	finally
	{
		Set-Location -Path $CurrentLocation
	}
}

Function Invoke-GitAdd
{
    Param(
	[Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitFolder,
    [Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$Path
	)
	$env:GIT_REDIRECT_STDERR = '2>&1'
	$CurrentLocation = Get-Location
	try
	{
		Set-Location -Path $GitFolder
		git add --force $Path
		if ($GitCommitAndPush) { git push --porcelain }
	}
	finally
	{
		Set-Location -Path $CurrentLocation
	}
}

Function Set-GitUserAndCredential
{
    Param(
	[Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitUserEmail,
	[Parameter(Mandatory=$true,ValueFromPipelinebyPropertyName=$True)]
    [String]$GitUserName,
	[Parameter(ValueFromPipelinebyPropertyName=$True)]
    [String]$GitCredentialUrl="",
	[Parameter(ValueFromPipelinebyPropertyName=$True)]
    [String]$GitCredentialPassword=""
	)
	$env:GIT_REDIRECT_STDERR = '2>&1'
	Write-Host "Set Git user email and name to $GitUserEmail and $GitUserName"
	git config --global user.email $GitUserEmail
	git config --global user.name $GitUserName
	if ($GiCredentialUrl -ne "" -and $GitCredentialPassword  -ne "") {
		Write-Host "Set Git credential to Windows Credential Manager for $GitCredentialUrl"
		$Result = cmdkey /generic:LegacyGeneric:target=$GitCredentialUrl /user:PersonalAccessToken /pass:$GitCredentialPassword
		Write-Host $Result
	}
}
